src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs

```diff
@@ -20,36 +20,37 @@ namespace Castle.Windsor.Extensions.DependencyInjection.Scope
 	using Castle.Core;
 	using Castle.MicroKernel;
 	using Castle.MicroKernel.Lifestyle.Scoped;
-	
+
 	internal class ExtensionContainerScope : ILifetimeScope, IDisposable
 	{
-		public static ExtensionContainerScope Current => current.Value;
+		internal static ExtensionContainerScope Current
+		{
+			get => current.Value;
+			set => current.Value = value;
+		}
 		public static string TransientMarker = "Transient";
-		protected static readonly AsyncLocal<ExtensionContainerScope> current = new AsyncLocal<ExtensionContainerScope>();
+		protected static readonly AsyncLocal<ExtensionContainerScope> current  = new AsyncLocal<ExtensionContainerScope>();
 		private readonly ExtensionContainerScope parent;
 		private readonly IScopeCache scopeCache;
 
-		protected ExtensionContainerScope(ExtensionContainerScope parent)
+		protected ExtensionContainerScope(ExtensionContainerScope parent, ExtensionContainerRootScope rootScope)
 		{
 			scopeCache = new ScopeCache();
-			if(parent == null)
-			{
-				this.parent = ExtensionContainerRootScope.RootScope;
-			}
-			else
-			{
-				this.parent = parent;
-			}
+			RootScope = rootScope ?? this as ExtensionContainerRootScope;
+			this.parent = parent;
 		}
 
-		public static ExtensionContainerScope BeginScope(ExtensionContainerScope parent)
+		internal ExtensionContainerRootScope RootScope { get; }
+
+		internal static ExtensionContainerScope BeginScope()
 		{
-			var scope = new ExtensionContainerScope(parent);
+			var parent = Current;
+			if (parent == null) throw new ArgumentNullException($"There is no parent scope to allow for BeginScope");
```

Comment:
No need for a `$` since there is no interpolation.

I wasn't saying the exception isn't needed, but that an `ArgumentNullException` isn't the right exception:
> The exception that is thrown when a null reference (Nothing in Visual Basic) is passed to a method that does not accept it as a valid argument.

Right now you are passing that string as `paramName`:
```
public ArgumentNullException (string paramName);
```

Maybe `InvalidOperationException` makes more sense.

Reply:
No need for a $ since there is no interpolation. --> This was a R# recommendation. I will remove